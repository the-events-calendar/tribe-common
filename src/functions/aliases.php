<?php

/**
 * Registers the library aliases redirecting calls to the `tad_DI52_`, non-namespaced, class format to the namespaced
 * classes.
 */

$aliases = [
	[ 'TEC\Common\lucatume\DI52\Container', 'tad_DI52_Container' ],
	[ 'TEC\Common\lucatume\DI52\ServiceProvider', 'tad_DI52_ServiceProvider' ]
];

foreach ( $aliases as [$class, $alias] ) {
	if ( ! class_exists( $alias ) ) {
		class_alias( $class, $alias );
	}
}