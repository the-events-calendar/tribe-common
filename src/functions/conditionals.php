<?php

/**
 * Determines if upsells should be hidden.
 *
 * @since 4.14.16
 *
 * @param string $slug Which upsell is this conditional for, if nothing passed it will apply to all.
 *
 * @return bool
 */
function tec_should_hide_upsell( string $slug = 'all' ): bool {
	$verify = static function( $needle, $haystack ) {
		// In all cases if true or false boolean we return that.
		if ( is_bool( $haystack ) ) {
			return $haystack;
		}

		if ( is_string( $haystack ) ) {
			// When all just return true to hide.
			if ( 'all' === $haystack ) {
				return true;
			}

			$truthy = tribe_is_truthy( $haystack );
			if ( $truthy ) {
				return $truthy;
			}
		}

		// Now allow multiple to be targeted as a string.
		$haystack = explode( '|', $haystack );

		// If the  `all` string is on the haystack
		if ( in_array( 'all', $haystack, true ) ) {
			return true;
		}

		return in_array( $needle, $haystack, true );
	};

	// If upsells have been manually hidden, respect that.
	if ( defined( 'TEC_HIDE_UPSELL' ) ) {
		return $verify( $slug, TEC_HIDE_UPSELL );
	}

	// If upsells have been manually hidden, respect that.
	if ( defined( 'TRIBE_HIDE_UPSELL' ) ) {
		return $verify( $slug, TRIBE_HIDE_UPSELL );
	}

	$env_var = getenv( 'TEC_HIDE_UPSELL' );
	if ( false !== $env_var ) {
		return $verify( $slug, $env_var );
	}

	/**
	 * Allows filtering of the Upsells for anything using Common.
	 *
	 * @since 4.14.16
	 *
	 * @param bool|string $hide Determines if Upsells are hidden.
	 * @param bool|string $slug Which slug we are testing against.
	 */
	$haystack = apply_filters( 'tec_should_hide_upsell', false, $slug );

	/**
	 * Allows filtering of the Upsells for anything using Common, for one specific slug.
	 *
	 * @since 4.14.16
	 *
	 * @param bool|string $hide Determines if Upsells are hidden.
	 * @param bool|string $slug Which slug we are testing against.
	 */
	$haystack = apply_filters( "tec_should_hide_upsell_{$slug}", $haystack, $slug );

	return $verify( $slug, $haystack );
}