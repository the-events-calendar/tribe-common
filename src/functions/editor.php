<?php
if ( ! function_exists( 'tec_is_full_site_editor' ) ) {
	/**
	 * Check if the current theme is a block theme.
	 *
	 * @since 4.14.18
	 *
	 * @return bool Whether the current theme is a block theme supporting full-site editing.
	 */
	function tec_is_full_site_editor() {
		if ( function_exists( 'wp_is_block_theme' ) ) {
			return (bool) wp_is_block_theme();
		} else if ( function_exists( 'gutenberg_is_fse_theme' ) ) {
			// This function has returned wp_is_block_theme since 2021/12 so this is just in case someone hasn't updated.
			return (bool) gutenberg_is_fse_theme();
		}

		return false;
	}
}

if ( ! function_exists( 'tec_using_classy_editor' ) ) {
	function tec_using_classy_editor(): bool {
		/**
		 * Filters whether the Classy experience is being used or not.
		 *
		 * @since TBD
		 *
		 * @param bool $using_classy_editor Whether the Classy experience is being used or not.
		 */
		$using_classy_editor = apply_filters( 'tec_using_classy_editor', false );

		return (bool) $using_classy_editor;
	}
}
