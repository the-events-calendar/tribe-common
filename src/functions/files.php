<?php
/**
 * Provides functions to seek and interact with files.
 *
 * @since 5.0.0
 */

if ( ! function_exists( 'tec_is_file_from_plugins' ) ) {
	/**
	 * Checks if a file is from one of the specified plugins.
	 *
	 * @since 5.0.0
	 *
	 * @param string $file            The path of the file to check.
	 * @param string ...$plugin_files A set of plugin main files to check, e.g. `the-events-calendar.php`.
	 *
	 * @return bool Whether the file is from one of the specified plugins.
	 */
	function tec_is_file_from_plugins( string $file, string ...$plugin_files ): bool {
		static $wp_active_and_valid_plugins = null;

		if ( empty( $wp_active_and_valid_plugins ) ) {
			// The list is expensive to generate, so we cache it.
			$wp_active_and_valid_plugins = wp_get_active_and_valid_plugins();
		}

		$plugin_dirs = array_map(
			'dirname',
			array_filter( $wp_active_and_valid_plugins, static function ( string $plugin ) use ( $plugin_files ): bool {
				return in_array( basename( $plugin ), $plugin_files, true );
			} )
		);

		foreach ( $plugin_dirs as $plugin_dir ) {
			if ( strpos( $file, $plugin_dir ) !== false ) {
				return true;
			}
		}

		return false;
	}
}
