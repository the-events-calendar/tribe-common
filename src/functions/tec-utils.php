<?php
/**
 * Utility functions that rely on tribe(). To be called after tribe() has been initialized.
 */

use TEC\Common\lucatume\DI52\ContainerException;

if ( ! function_exists( 'tec_copy_to_clipboard_button' ) ) {
	/**
	 * Output a button to copy the content of an element to the clipboard.
	 *
	 * @since 6.0.3
	 *
	 * @param string $content_to_copy The content to copy to the clipboard.
	 * @param bool   $output_button   Whether to output the button or just the target element.
	 * @param string $aria_label      The aria-label attribute for the button.
	 *
	 * @return string
	 */
	function tec_copy_to_clipboard_button( string $content_to_copy, bool $output_button = true, string $aria_label = '' ): string {
		$cache_key = 'tec_copy_to_clipboard_counter';
		$counter   = tribe( 'cache' )->get( $cache_key, '', 1 );

		$target        = 'tec-copy-text-target-' . $counter;
		$notice_target = 'tec-copy-to-clipboard-notice-content-' . $counter;

		$aria_label = $aria_label ? $aria_label : __( 'Copy to clipboard', 'the-events-calendar' );
		tribe( 'cache' )->set( $cache_key, ++$counter, Tribe__Cache::NON_PERSISTENT );
		if ( $output_button ) :
			?>
			<a
				title="<?php echo esc_attr( $aria_label ); ?>"
				href="javascript:void(0)"
				aria-label="<?php echo esc_attr( $aria_label ); ?>"
				aria-describedby="<?php echo esc_attr( $notice_target ); ?>"
				data-clipboard-action="copy"
				data-notice-target=".<?php echo esc_attr( $notice_target ); ?>"
				data-clipboard-target=".<?php echo esc_attr( $target ); ?>"
				class="tec-copy-to-clipboard tribe-dashicons"
				role="button"
			>
				<input type="text" readonly value="<?php echo esc_attr( $content_to_copy ); ?>" aria-hidden="true" />
				<span class="dashicons dashicons-admin-page" aria-hidden="true"></span>
			</a>
		<?php
		endif;
		?>
		<span class="screen-reader-text <?php echo esc_attr( $target ); ?>"><?php echo esc_html( $content_to_copy ); ?></span>
		<div class="tec-copy-to-clipboard-notice">
			<div class="tec-copy-to-clipboard-notice-content <?php echo esc_attr( $notice_target ); ?>" aria-live="polite">
			</div>
		</div>
		<?php

		// When they want to print the button outside of this function they need to be aware of the target.
		return $target;
	}
}

if ( ! function_exists( 'tribe_register_plugin' ) ) {
	/**
	 * Checks if this plugin has permission to run, if not it notifies the admin
	 *
	 * @param string $file_path    Full file path to the base plugin file.
	 * @param string $main_class   The Main/base class for this plugin.
	 * @param string $version      The version.
	 * @param array  $classes_req  Any Main class files/tribe plugins required for this to run.
	 * @param array  $dependencies an array of dependencies to check.
	 */
	function tribe_register_plugin( $file_path, $main_class, $version, $classes_req = [], $dependencies = [] ) {
		$tribe_dependency = tribe( Tribe__Dependency::class );
		$tribe_dependency->register_plugin( $file_path, $main_class, $version, $classes_req, $dependencies );
	}
}

if ( ! function_exists( 'tribe_get_class_instance' ) ) {
	/**
	 * Gets the class instance / Tribe Container from the passed object or string.
	 *
	 * @since 4.10.0
	 *
	 * @see   \TEC\Common\lucatume\DI52\Builders\ValueBuilder\App::isBound()
	 * @see   \tribe()
	 *
	 * @param string|object $class The plugin class' singleton name, class name, or instance.
	 *
	 * @return mixed|object|Tribe__Container|null Null if not found, else the result from tribe().
	 */
	function tribe_get_class_instance( $class ) {
		if ( is_object( $class ) ) {
			return $class;
		}

		if ( ! is_string( $class ) ) {
			return null;
		}

		// Check if class exists and has instance getter method.
		if ( class_exists( $class ) ) {
			if ( method_exists( $class, 'instance' ) ) {
				return $class::instance();
			}

			if ( method_exists( $class, 'get_instance' ) ) {
				return $class::get_instance();
			}
		}

		try {
			return tribe()->has( $class ) ? tribe()->get( $class ) : null;
		} catch ( ContainerException $exception ) {
			return null;
		}
	}
}

/**
 * Get the next increment of a cached incremental value.
 *
 * @since 4.14.7
 *
 * @param string $key Cache key for the incrementor.
 * @param string $expiration_trigger The trigger that causes the cache key to expire.
 * @param int    $default The default value of the incrementor.
 *
 * @return int
 **/
function tribe_get_next_cached_increment( $key, $expiration_trigger = '', $default = 0 ) {
	$cache = tribe( 'cache' );
	$value = (int) $cache->get( $key, $expiration_trigger, $default );
	++$value;
	$cache->set( $key, $value, \Tribe__Cache::NON_PERSISTENT, $expiration_trigger );

	return $value;
}

if ( ! function_exists( 'tribe_check_plugin' ) ) {
	/**
	 * Checks if this plugin has permission to run, if not it notifies the admin
	 *
	 * @since 4.9
	 *
	 * @param string $main_class The Main/base class for this plugin.
	 *
	 * @return bool Indicates if plugin should continue initialization
	 */
	function tribe_check_plugin( $main_class ) {

		$tribe_dependency = tribe( Tribe__Dependency::class );

		return $tribe_dependency->check_plugin( $main_class );
	}
}
