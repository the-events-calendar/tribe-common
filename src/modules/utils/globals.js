/**
 * @todo: handle globals in a better way
 */

export const get = ( key, defaultValue ) => window[ key ] || defaultValue;
export const google = () => get( 'google' );
export const wpApi = wp.api;
export const wpApiRequest = wp.apiRequest;
export const wpComponents = wp.components;
export const wpData = wp.data;
export const wpEditor = wp.blockEditor || wp.editor;
export const wpHooks = wp.hooks;

export const wpDataSelectCoreEditor = () => wpData.select( 'core/block-editor' ) ||
	wpData.select( 'core/editor' );

export const IconButton = wpComponents.Button || wpComponents.IconButton;

// Localized Config
export const config = () => get( 'tribe_editor_config', {} );

// Common
export const common = () => config().common || {};
export const adminUrl = () => common().adminUrl || '';
export const rest = () => common().rest || {};
export const restNonce = () => rest().nonce || {};
export const dateSettings = () => common().dateSettings || {};
export const editorConstants = () => common().constants || {};
export const list = () => ( {
	countries: common().countries || {},
	us_states: common().usStates || {},
} );

// TEC
export const tec = () => config().events || {};
export const editor = () => tec().editor || {};
export const settings = () => tec().settings || {};
export const mapsAPI = () => tec().googleMap || {};
export const priceSettings = () => tec().priceSettings || {};
export const tecDateSettings = () => tec().dateSettings || {};
export const timezoneHtml = () => tec().timezoneHTML || '';
export const defaultTimes = () => tec().defaultTimes || {};
export const timezone = () => tec().timeZone || {};

// PRO
export const pro = () => config().eventsPRO || {};
export const editorDefaults = () => pro().defaults || {};

// Tickets
export const tickets = () => config().tickets || {};

// Tickets Plus
export const ticketsPlus = () => config().ticketsPlus || {};
export const iacVars = () => ticketsPlus().iacVars || {};

// Post Objects
export const postObjects = () => config().post_objects || {};
export const post = () => config().post || {};

// Blocks
export const blocks = () => config().blocks || {};
