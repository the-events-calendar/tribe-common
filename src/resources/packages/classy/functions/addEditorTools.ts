import { __ } from '@wordpress/i18n';
import { registerPlugin } from '@wordpress/plugins';

/**
 * CSS class names for the editor tools.
 *
 * @since TBD
 */
const EDITOR_TOOL_CLASSES = {
	EDITOR_DOCUMENT_TOOLS: 'editor-document-tools',
	EDITOR_DOCUMENT_TOOLS_LEFT: 'editor-document-tools__left',
	TEC_EDITOR_TOOL: 'tec-editor-tool',
	TEC_EDITOR_TOOL_PREVIEW: 'tec-editor-tool--preview',
	BUTTON: 'button',
	DASHICONS_VISIBILITY: 'dashicons dashicons-visibility',
} as const;

/**
 * Data attributes for the editor tools.
 *
 * @since TBD
 */
const DATA_ATTRIBUTES = {
	TOOLBAR_ITEM: 'data-toolbar-item',
} as const;

/**
 * Plugin registration name.
 *
 * @since TBD
 */
const PLUGIN_NAME = 'tec-editor-tools';

/**
 * Adds the Editor tools inserting them among the Block Editor tools.
 *
 * This function is weird-looking to work around the lack of an official API to add/remove Editor tools.
 * The function registers a plugin to be called when the toolbar is rendered; then calls a function that looks like
 * a React Component (EditorTools) that will not actually render a component, but will use the call to insert a button
 * in the Editor tools.
 *
 * @since TBD
 *
 * @param {(this: GlobalEventHandlers, ev: MouseEvent) => void} onClick The click event to handle on press of the added
 *     button.
 * @param {Document|null} document The document to work on, or `null` to work on `window.document`.
 *
 * @return {void} The Editor tools are added to the toolbar.
 */
export function addEditorTools(
	onClick: ( this: GlobalEventHandlers, ev: MouseEvent ) => void,
	document: Document | null = null
): void {
	document = document || window.document;

	// 2. The function _should_ render a React component, but won't. It will insert a button.
	function EditorTools() {
		const editorDocumentTools = document.querySelector(
			`.${ EDITOR_TOOL_CLASSES.EDITOR_DOCUMENT_TOOLS } .${ EDITOR_TOOL_CLASSES.EDITOR_DOCUMENT_TOOLS_LEFT }`
		);
		const previewButton = document.querySelector( `.${ EDITOR_TOOL_CLASSES.TEC_EDITOR_TOOL_PREVIEW }` );

		if ( editorDocumentTools && previewButton === null ) {
			const previewButton = document.createElement( 'button' );
			previewButton.classList.add(
				EDITOR_TOOL_CLASSES.TEC_EDITOR_TOOL,
				EDITOR_TOOL_CLASSES.TEC_EDITOR_TOOL_PREVIEW,
				EDITOR_TOOL_CLASSES.BUTTON
			);
			previewButton.type = 'button';
			previewButton.setAttribute( DATA_ATTRIBUTES.TOOLBAR_ITEM, 'true' );
			previewButton.innerHTML = `<span class="${ EDITOR_TOOL_CLASSES.DASHICONS_VISIBILITY }"></span>${ __(
				'Visual',
				'tribe-common'
			) }`;
			editorDocumentTools.append( previewButton );
			previewButton.onclick = onClick;
		}

		// 3. Do not actually render any component.
		return null;
	}

	// 1. When the time comes for plugin registration, it's the moment to insert the button.
	registerPlugin( PLUGIN_NAME, {
		render: EditorTools,
	} );
}
