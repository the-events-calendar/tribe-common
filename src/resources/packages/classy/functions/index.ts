export * from './dateUtils';
export { getDefaultRegistry, setDefaultRegistry } from './getDefaultRegistry';
export { sortOptionsForDisplay } from './sortOptionsForDisplay';
