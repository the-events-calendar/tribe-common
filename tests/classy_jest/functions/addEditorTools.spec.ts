import { addEditorTools } from '../../../src/resources/packages/classy/functions/addEditorTools';
import { afterEach, describe, it, expect, jest } from '@jest/globals';
import { registerPlugin } from '@wordpress/plugins';

jest.mock( '@wordpress/plugins', () => ( {
	registerPlugin: jest.fn( ( name: string, settings ) => null ),
} ) );

/**
 * Tests for the addEditorTools function.
 *
 * Note: In production, this function is only called when the `classy_enable_visual=1`
 * query parameter is present in the URL (checked in index.ts). These tests focus on
 * the function's behavior when called, not the conditional invocation logic.
 */
describe( 'addEditorTools', () => {
	afterEach( () => {
		jest.resetAllMocks();
		jest.restoreAllMocks();
	} );

	it( 'should register plugin with correct name and render function', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                    <div class="editor-document-tools__left"></div>
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		expect( registerPlugin ).toHaveBeenCalledWith( 'tec-editor-tools', {
			render: expect.any( Function ),
		} );
	} );

	it( 'should add editor tools button to the toolbar', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                    <div class="editor-document-tools__left"></div>
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		expect( onClick ).not.toHaveBeenCalled();
		expect( mockDocument.querySelectorAll( '.editor-document-tools__left .tec-editor-tool' ).length ).toBe( 0 );
		expect( registerPlugin ).toHaveBeenCalledWith( 'tec-editor-tools', {
			render: expect.any( Function ),
		} );

		const registerPluginMock = registerPlugin as jest.Mock;
		const registerPluginMockCallSettings = registerPluginMock.mock.calls[ 0 ][ 1 ] as { render: Function };
		const renderEditorTools = registerPluginMockCallSettings.render;

		const renderedEditorTools = renderEditorTools();

		expect( renderedEditorTools ).toBeNull();
		expect( mockDocument.querySelectorAll( '.editor-document-tools__left .tec-editor-tool' ).length ).toBe( 1 );
		expect(
			mockDocument.querySelector( '.editor-document-tools__left .tec-editor-tool' ).outerHTML
		).toMatchSnapshot();
	} );

	it( 'should not add duplicate buttons on multiple render calls', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                    <div class="editor-document-tools__left"></div>
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		const registerPluginMock = registerPlugin as jest.Mock;
		const registerPluginMockCallSettings = registerPluginMock.mock.calls[ 0 ][ 1 ] as { render: Function };
		const renderEditorTools = registerPluginMockCallSettings.render;

		renderEditorTools();

		expect( mockDocument.querySelectorAll( '.editor-document-tools__left .tec-editor-tool' ).length ).toBe( 1 );

		const secondRenderedEditorTools = renderEditorTools();

		expect( secondRenderedEditorTools ).toBeNull();
		expect( mockDocument.querySelectorAll( '.editor-document-tools__left .tec-editor-tool' ).length ).toBe( 1 );
	} );

	it( 'should call onClick handler when button is clicked', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                    <div class="editor-document-tools__left"></div>
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		const registerPluginMock = registerPlugin as jest.Mock;
		const registerPluginMockCallSettings = registerPluginMock.mock.calls[ 0 ][ 1 ] as { render: Function };
		const renderEditorTools = registerPluginMockCallSettings.render;

		renderEditorTools();

		const button = mockDocument.querySelector( '.editor-document-tools__left .tec-editor-tool' ) as HTMLElement;
		button.click();
		expect( onClick ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'should not add any button if the document does not have the editor tools container', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		expect( onClick ).not.toHaveBeenCalled();
		expect( mockDocument.querySelectorAll( '.editor-document-tools__left .tec-editor-tool' ).length ).toBe( 0 );
		expect( registerPlugin ).toHaveBeenCalledWith( 'tec-editor-tools', {
			render: expect.any( Function ),
		} );

		const registerPluginMock = registerPlugin as jest.Mock;
		const registerPluginMockCallSettings = registerPluginMock.mock.calls[ 0 ][ 1 ] as { render: Function };
		const renderEditorTools = registerPluginMockCallSettings.render;

		const renderedEditorTools = renderEditorTools();

		expect( renderedEditorTools ).toBeNull();
		expect( mockDocument.querySelectorAll( '.editor-document-tools .tec-editor-tool' ).length ).toBe( 0 );
	} );

	it( 'should use window.document when no document is provided', () => {
		const onClick = jest.fn();

		addEditorTools( onClick );

		expect( registerPlugin ).toHaveBeenCalledWith( 'tec-editor-tools', {
			render: expect.any( Function ),
		} );
	} );

	it( 'should add button with correct CSS classes', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                    <div class="editor-document-tools__left"></div>
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		const registerPluginMock = registerPlugin as jest.Mock;
		const registerPluginMockCallSettings = registerPluginMock.mock.calls[ 0 ][ 1 ] as { render: Function };
		const renderEditorTools = registerPluginMockCallSettings.render;

		renderEditorTools();

		const button = mockDocument.querySelector( '.editor-document-tools__left .tec-editor-tool' ) as HTMLElement;

		expect( button.classList.contains( 'tec-editor-tool' ) ).toBe( true );
		expect( button.classList.contains( 'tec-editor-tool--preview' ) ).toBe( true );
		expect( button.classList.contains( 'button' ) ).toBe( true );
		expect( button.getAttribute( 'type' ) ).toBe( 'button' );
		expect( button.getAttribute( 'data-toolbar-item' ) ).toBe( 'true' );
	} );

	it( 'should add button with visibility icon and text', () => {
		const mockDocument = new DOMParser().parseFromString(
			`<html><body>
                <div class="editor-document-tools">
                    <div class="editor-document-tools__left"></div>
                </div>
            </body></html>`,
			'text/html'
		);
		const onClick = jest.fn();

		addEditorTools( onClick, mockDocument );

		const registerPluginMock = registerPlugin as jest.Mock;
		const registerPluginMockCallSettings = registerPluginMock.mock.calls[ 0 ][ 1 ] as { render: Function };
		const renderEditorTools = registerPluginMockCallSettings.render;

		renderEditorTools();

		const button = mockDocument.querySelector( '.editor-document-tools__left .tec-editor-tool' ) as HTMLElement;

		expect( button.innerHTML ).toContain( 'dashicons-visibility' );
		expect( button.innerHTML ).toContain( 'Visual' );
	} );
} );
