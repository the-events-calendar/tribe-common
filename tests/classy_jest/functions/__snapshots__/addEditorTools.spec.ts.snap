// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addEditorTools should add editor tools button to the toolbar 1`] = `"<button class="tec-editor-tool tec-editor-tool--preview button" type="button" data-toolbar-item="true"><span class="dashicons dashicons-visibility"></span>Visual</button>"`;
